#  Analytical Decision Making  - Optimization
### By Mei Yang 

## 1. Sensitivity Analyses

### Background
Content publishers such as The New York Times, The Washington Post and The Wall Street Journal generate revenue by using display advertisements. The Washington Post's website contains several different sections including Sports and National. The number of views each section gets per day can be estimated by analyzing historical data. Assume that the Sports section gets six million views per day and the National section get five million views per day. Assume four companies, GEICO, Delta, T-Mobile and Capital One, wish to advertise on the Sports and National sections of the Washington Post and they contract directly with the newspaper. For each company, the contract specifies the number of times its display ads are shown in these two sections. The contracts sometimes also specify a total number of page views that can originate from any section of the newspaper. The page views promised by The Washington Post to each advertiser are summarized in the left table below.  Assume that the contract also specifies that The Washington Post receives $2.40 per click-through from each of the four companies. However, not every page view leads to a click. If every 1000 views leads to 5 clicks, the click-through rate is 0.005. Newspapers use historical data and tracking technologies to determine click-through rates. Assume that the relevant click-through rates are given in the right table below. 
![54323](https://user-images.githubusercontent.com/32555702/39460899-ec280eea-4cbb-11e8-9425-f6e56db049e5.JPG)

### 1. Write out the optimization problem. Make sure to detail the decision variables, the objective function and the constraints. 


---


**Decision variables:**
The decision variables are the numbers of impressions for the four companies in two different sections of the Washing Post.
Let i = 1, 2, 3, 4 be the index of the four companies, and j = 1, 2 be the index of the two sections ----
![capture](https://user-images.githubusercontent.com/32555702/39460900-ec47e530-4cbb-11e8-8b09-328ece2bb627.JPG)

**Objective function:**    
The objective of the Washington Post is to maximize advertising revenues, which is determined by the $2.4 per click charged to the four companies and the click-through rates. Let  Kij  be the click through rates of the four companies, in which  i  represents the company and j represent the section ----
![54](https://user-images.githubusercontent.com/32555702/39460896-ebe69b04-4cbb-11e8-8368-170041cfd101.JPG)

The objective function is

\begin{equation} Objective=2.4\times \sum_{i=1}^4\sum_{j=1}^{2}  \kappa_{ij}\times x_{ij} \end{equation}

**Constraints:**
Given that the Sports section gets six million views per day and the National section get five million views per day, and the contract of minimum number of impressions required, we can list the constraints:

Geico:

    x11>=2,000,000

    x12>=1,000,000



Delta:

    x22>=1,000,000

    x21+x22=2,000,000



T-Mobile:

    x31>=1,000,000

    x32>=1,000,000

    x31+x32=3,000,000



Capital One:

    x41+x42=2,000,000



Sports:

    x11+x21+x31+x41<=6000000



National:

    x12+x22+x32+x42<=5000000


Non-negativity:

    xij >= 0



In [3]:
# install packages for convex optimization
!pip install cvxopt
!pip install cvxpy

    100% |████████████████████████████████| 11.5MB 3.9MB/s 
    100% |████████████████████████████████| 163kB 7.2MB/s 
    100% |████████████████████████████████| 122kB 8.9MB/s 
    100% |████████████████████████████████| 143kB 7.4MB/s 
    100% |████████████████████████████████| 1.5MB 11.1MB/s 
    100% |████████████████████████████████| 51kB 19.3MB/s 
    100% |████████████████████████████████| 696kB 8.1MB/s 
    100% |████████████████████████████████| 71kB 8.0MB/s 
  Running setup.py bdist_wheel for cvxpy ... - \ | / - \ | done
  Stored in directory: /content/.cache/pip/wheels/2d/26/ba/f0a0388a7ab98124caf88ea8b532b81cf4acb1b6a4a990e3b4
  Running setup.py bdist_wheel for ecos ... - \ | / - \ | / - \ | / -

 \ | / - \ | / - \ | done
  Stored in directory: /content/.cache/pip/wheels/50/91/1b/568de3c087b3399b03d130e71b1fd048ec072c45f72b6b6e9a
  Running setup.py bdist_wheel for scs ... - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ | / - \ done
  Stored in directory: /content/.cache/pip/wheels/ff/f0/aa/530ccd478d7d9900b4e9ef5bc5a39e895ce110bed3d3ac653e
  Running setup.py bdist_wheel for multiprocess ... - \ | / done
  Stored in directory: /content/.cache/pip/wheels/85/44/92/7635fa709a1e88a4dca84eba3d1566fc0e2b001d4d84f40d35
  Running setup.py bdist_wheel for fastcache ... - \ done
  Stored in directory: /content/.cache/pip/wheels/b7/90/c0/da92ac52d188d9ebca577044e89a14d0e6ff333c1bcd1ebc14
  Running setup.py bdist_wheel for toolz ... - \ done
  Stored in directory: /content/.cache/pip/wheels/f4/0c/f6/ce6b2d1aa459ee97cc3c0f82236302bd62d89c86c700219463
  Running

In [0]:
# import packages
import math
import numpy as np
import cvxpy as cvx


# Define the Variables
x11 = cvx.Int()# Geico Sports
x12 = cvx.Int()# Geico National
x21 = cvx.Int()# Delta Sports
x22 = cvx.Int()# Delta National
x31 = cvx.Int()# T-Mobile Sports
x32 = cvx.Int()# T-Mobile National
x41 = cvx.Int()# Capital One Sports
x42 = cvx.Int()# Capital One National

# Define the Constraints
c1= x11>= 2000000 
c2= x12>= 1000000 
c3=x22>= 1000000 
c4= x21+x22== 2000000 
c5=x31>=1000000
c6=x32>=1000000
c7=x31+x32==3000000
c8=x41+x42==2000000
c9=x11+x21+x31+x41<=6000000
c10=x12+x22+x32+x42<=5000000
c11=0<=x21
c12=0<=x41
c13=0<=x42

con=[c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13]


# Define the objective function
k11=2.5/100
k12=0.8/100
k21=2.0/100
k22=1.0/100
k31=1.0/100
k32=3.0/100
k41=1.5/100
k42=2.0/100

OF=(k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32+k41*x41+k42*x42)*2.4
objective = cvx.Maximize(OF)


#solving
prob = cvx.Problem(objective, con)
result = prob.solve()


### 2.	What is the optimal solution? Give the values of the decision variables and the optimal objective function value. 


---


Optimal Objective Function Value - 

    $523199.9999984559

Impression Allocation in the Sport Section - 

    GEICO: 2999999.999998966
    DELTA: 999999.999968759
    T-MOBILE: 1000000.0000099222
    CAPITAL ONE: 1000000.0000220642

Impression Allocation in the National Section - 

    GEICO: 1000000.0000004534
    DELTA: 1000000.0000312413
    T-MOBILE: 1999999.9999900782
    CAPITAL ONE: 999999.9999779359


In [21]:
print('The Optimal Objective Function Value Is:')
print("$", prob.value)

print('Allocation in the Sport Section')
print ("GEICO:", x11.value)
print ("DELTA:", x21.value)
print ("T-MOBILE:", x31.value)
print ("CAPITAL ONE:", x41.value)

print('Allocation in the National Section')
print ("GEICO:", x12.value)
print ("DELTA:", x22.value)
print ("T-MOBILE:", x32.value)
print ("CAPITAL ONE:", x42.value)

The Optimal Objective Function Value Is:
$ 523199.9999984559
Allocation in the Sport Section
GEICO: 2999999.999998966
DELTA: 999999.999968759
T-MOBILE: 1000000.0000099222
CAPITAL ONE: 1000000.0000220642
Allocation in the National Section
GEICO: 1000000.0000004534
DELTA: 1000000.0000312413
T-MOBILE: 1999999.9999900782
CAPITAL ONE: 999999.9999779359


### 3.	Conduct sensitivity analyses of the optimal solution to different click-through rates as given in the table below. Do one cell at time if you want, e.g., sensitivity of the solution to different CTRs for GEICO in the Sports section, then in the National section; then do the same for each advertiser. 
![1](https://user-images.githubusercontent.com/32555702/39461144-a2d6e110-4cbd-11e8-9c64-7c8741eeb9bf.JPG)


In [0]:
import pandas as pd

# Set range for sensitivity analysis with a difference of 0.0001
r11 = np.arange(0.02,0.03, 0.0001)   
r21 = np.arange(0.015,0.025, 0.0001) 
r31 = np.arange(0.005,0.015, 0.0001)
r41 = np.arange(0.015,0.020, 0.0001) 

r21 = np.arange(0.001,0.011, 0.0001) 
r22 = np.arange(0.005,0.015, 0.0001)
r32 = np.arange(0.025,0.035, 0.0001)  
r42 = np.arange(0.015,0.025, 0.0001) 

In [40]:
# Geico on Sports
r11_list = []

# def optimization function
def opt(n):
    OF_r11 = ((n*x11+k12*x12+k21*x21+k22*x22+k31*x31+k32*x32+k41*x41+k42*x42)*2.4)
    obj_r11 = cvx.Maximize(OF_r11)
    prob_r11 = cvx.Problem(obj_r11, con)
    return prob_r11.solve()

# sensitivity analysis
for n in (r11):
    r11_list.append(opt(n))
    
df_r11 = pd.DataFrame(data=[r11,r11_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r11.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for Geico on Sports:")
df_r11.max()

Sensitivity Analysis Result for Geico on Sports:


Click Through Rate         0.029900
Optimal Revenue       558479.999997
dtype: float64

In [45]:
# Geico on National
r21_list = []

# def optimization function
def opt(n):
    OF_r21 = ((k11*x11+n*x12+k21*x21+k22*x22+k31*x31+k32*x32+k41*x41+k42*x42)*2.4)
    obj_r21 = cvx.Maximize(OF_r21)
    prob_r21 = cvx.Problem(obj_r21, con)
    return prob_r21.solve()

# sensitivity analysis
for n in (r21):
    r21_list.append(opt(n))
    
df_r21 = pd.DataFrame(data=[r21,r21_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r21.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for Geico on National:")
df_r21.max()

Sensitivity Analysis Result for Geico on National:


Click Through Rate         0.010900
Optimal Revenue       530159.999999
dtype: float64

In [58]:
# Delta on Sports
r21_list = []

# def optimization function
def opt(n):
    OF_r21 = ((k11*x11+k12*x12+n*x21+k22*x22+k31*x31+k32*x32+k41*x41+k42*x42)*2.4)
    obj_r21 = cvx.Maximize(OF_r21)
    prob_r21 = cvx.Problem(obj_r21, con)
    return prob_r21.solve()

# sensitivity analysis
for n in (r21):
    r21_list.append(opt(n))
    
df_r21 = pd.DataFrame(data=[r21,r21_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r21.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for Delta on Sports:")
df_r21.max()

Sensitivity Analysis Result for Delta on Sports:


Click Through Rate         0.0109
Optimal Revenue       501360.0000
dtype: float64

In [59]:
# Delta on National
r22_list = []

# def optimization function
def opt(n):
    OF_r22 =  ((k11*x11+k12*x12+k21*x21+n*x22+k31*x31+k32*x32+k41*x41+k42*x42)*2.4)
    obj_r22 = cvx.Maximize(OF_r22)
    prob_r22 = cvx.Problem(obj_r22, con)
    return prob_r22.solve()

# sensitivity analysis
for n in (r22):
    r22_list.append(opt(n))
    
df_r22 = pd.DataFrame(data=[r22,r22_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r22.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for Delta on National:")
df_r22.max()

Sensitivity Analysis Result for Delta on National:


Click Through Rate         0.0149
Optimal Revenue       534960.0000
dtype: float64

In [60]:
# T-Mobile on Sports
r31_list = []

# def optimization function
def opt(n):
    OF_r31 = ((k11*x11+k12*x12+k21*x21+k22*x22+n*x31+k32*x32+k41*x41+k42*x42)*2.4)
    obj_r31 = cvx.Maximize(OF_r31)
    prob_r31 = cvx.Problem(obj_r31, con)
    return prob_r31.solve()

# sensitivity analysis
for n in (r31):
    r31_list.append(opt(n))
    
df_r31 = pd.DataFrame(data=[r31,r31_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r31.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for T-Mobile on Sports:")
df_r31.max()

Sensitivity Analysis Result for T-Mobile on Sports:


Click Through Rate         0.014900
Optimal Revenue       534959.999999
dtype: float64

In [61]:
#T-Mobile on National
r32_list = []

# def optimization function
def opt(n):
    OF_r32 = ((k11*x11+k12*x12+k21*x21+k22*x22+k31*x31+n*x32+k41*x41+k42*x42)*2.4)
    obj_r32 = cvx.Maximize(OF_r32)
    prob_r32 = cvx.Problem(obj_r32, con)
    return prob_r32.solve()

# sensitivity analysis
for n in (r32):
    r32_list.append(opt(n))
    
df_r32 = pd.DataFrame(data=[r32,r32_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r32.columns = names

# find CTR with the optimal revenue
print("Sensit#ivity Analysis Result for T-Mobile on National:")
df_r32.max()

Sensit#ivity Analysis Result for T-Mobile on National:


Click Through Rate         0.035000
Optimal Revenue       547199.999999
dtype: float64

In [62]:
# Capital One on Sports
r41_list = []

# def optimization function
def opt(n):
    OF_r41 = ((k11*x11+k12*x12+k21*x21+k22*x22+k31*x31+k32*x32+n*x41+k42*x42)*2.4)
    obj_r41 = cvx.Maximize(OF_r41)
    prob_r41 = cvx.Problem(obj_r41, con)
    return prob_r41.solve()

# sensitivity analysis
for n in (r41):
    r41_list.append(opt(n))
    
df_r41 = pd.DataFrame(data=[r41,r41_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r41.columns = names

# find CTR with the optimal revenue
print("Sensitivity Analysis Result for T-Mobile on Sports:")
df_r41.max()

Sensitivity Analysis Result for T-Mobile on Sports:


Click Through Rate         0.020000
Optimal Revenue       535199.999999
dtype: float64

In [55]:
# Capital One on National
r42_list = []

# def optimization function
def opt(n):
    OF_r42 = ((k11*x11+k12*x12+k21*x21+k22*x22+k31*x31+k32*x32+k41*x41+n*x42)*2.4)
    obj_r42 = cvx.Maximize(OF_r42)
    prob_r42 = cvx.Problem(obj_r42, con)
    return prob_r42.solve()

# sensitivity analysis
for n in (r42):
    r42_list.append(opt(n))
    
df_r42 = pd.DataFrame(data=[r42,r42_list]).transpose()
names = ['Click Through Rate', 'Optimal Revenue']
df_r42.columns = names

# find CTR with the optimal revenue
print("Sensit#ivity Analysis Result for Capital One on National:")
df_r42.max()

Sensit#ivity Analysis Result for Capital One on National:


Click Through Rate         0.025
Optimal Revenue       535200.000
dtype: float64

### 4. Write no more than one paragraph to describe the results of your analyses. Tables or graphs can be provided as supplemental material.

---

According the results of sensitivity analyses, click through rates that can bring optimal strategy are all found at the higher end of the given ranges. Therefore, we can conclude that increasing click through rate is a good way to increase revenue. However, comparing the difference between revenue increased by increasing click through rate, the increase in revenue isn't very rapid. So when investing in increasing click through rate, the Washing Post might want to pay attention the costs of increasing click through rate and ROI.

## 2: Operations Excellence 

### Background
You operate two plants, i.e., A and B. Each plant makes two products, “standard” and “deluxe”. A unit of standard gives a profit contribution of 10 dollars, while a unit of deluxe gives a profit contribution of 15 dollars. 
 
Each plant uses two processes, grinding and polishing, for producing its products. Plant A has a grinding capacity of 80 hours per week and polishing capacity of 60 hours per week. For plant B, these capacities are 60 and 75 hours per week, respectively. 
 
The grinding and polishing times in hours for a unit of each type of product in each factory are given in the tables below. 

| Plant A |  Standard  |  Deluxe          
| :---:|:---: | :---:
|   Grinding  |   4   |   2
|   Polishing |   2   |   5


| Plant B |  Standard  |  Deluxe          
| :---:|:---: | :---:
|   Grinding  |   5   |   3
|   Polishing |   5   |   6


It is possible, for example, that plant B has older machines than plant A, resulting in higher unit processing times. In addition, each unit of each product uses 4 kg of a raw material, which we refer to as raw. The company has 120 kg of raw available per week. To start with, we will assume that plant A is allocated 75 kg of raw per week and plant B the remaining 45 kg per week. Each plant can build a very simple linear programming model to maximize its profit contribution.  


### 1.	Write the optimization models for each plant, i.e., make sure to detail the decision variables, the objective function and the constraints. 


---


**Decision Variable:** 
Let i be the index of plants, i = a represent plant A and i = b represents plant; let j be the index of products, j = 1 represent "Standard" and j = 2 represents "Deluxe".
The decision variables, Xij, will be the number of units each product produced by each plant ----

| Plant|  Standard  |  Deluxe          
| :---:|:---: | :---:
|   A  |   xa1   |   xa2
|   B  |   xb1   |   xb2


**Objective Function:**
The objective is for each plant to maximize its profit contribution, so we hace objective functions:

     Plant A: max 10Xa1+15Xa2 

     Plant B: max 10Xb1+15Xb2 

**Constraints:**
According to grinding and polishing capacities of each plant and that plant A is allocated 75 kg of raw per week and plant B the remaining 45 kg per week, the constraints are:

Plant A: 

    4Xa1+4Xa2<=75

    4Xa1+2Xa2<=80

    2Xa1+5Xa2<=60

    Xa1,Xa2>=0

        
Plant B: 

    4Xb1+4Xb2<=45

    5Xb1+3Xb2<=60

    5Xb1+6Xb2<=75

    Xb1,Xb2>=0

### 2.	Provide the optimal solutions and profit levels for each factory. Give the values of the decision variables and the optimal objective function value. 


---


I listed 3 ways to solve this problem, and solution 3 worked the best. The answers are:


    PLant A should produce 10.0 units of standard product and 8.0 units of deluxe product. 
    The maxmized profit of Plant A is $ 220.0.
    
    PLant B should produce 0.0 units of standard product and 11.0 units of deluxe product. 
    The maxmized profit of Plant B is $ 165.0.

#### *Solution 1*

In [29]:
# try solve the porblem with CVXOPT solver
import cvxopt
from cvxopt import matrix, solvers
  
# plant A
A = matrix([ [4.0, 4.0, 2.0,-1.0,0.0], [4.0, 2.0, 5.0, 0.0,-1.0] ])
b = matrix([ 75.0, 80.0, 60.0, 0.0, 0.0 ])
c = matrix([ -10.0, -15.0 ])
sol=solvers.lp(c,A,b)
print(sol['x'])
print(-sol['primal objective'])
print(-numpy.matmul(numpy.transpose(c),sol['x']))

     pcost       dcost       gap    pres   dres   k/t
 0: -2.3150e+02 -5.3116e+02  1e+02  1e-01  1e+00  1e+00
 1: -2.2424e+02 -2.4880e+02  7e+00  1e-02  9e-02  4e-01
 2: -2.2499e+02 -2.2530e+02  8e-02  1e-04  1e-03  6e-03
 3: -2.2500e+02 -2.2500e+02  8e-04  1e-06  1e-05  6e-05
 4: -2.2500e+02 -2.2500e+02  8e-06  1e-08  1e-07  6e-07
 5: -2.2500e+02 -2.2500e+02  8e-08  1e-10  1e-09  6e-09
Optimal solution found.
[ 1.13e+01]
[ 7.50e+00]

224.99999998590417
[[224.99999999]]


In [30]:
# plant B
A = matrix([ [4.0, 5.0, 5.0,-1.0,0.0], [4.0, 3.0, .0, 0.0,-1.0] ])
b = matrix([ 45.0, 60.0, 75.0, 0.0, 0.0 ])
c = matrix([ -10.0, -15.0 ])
sol=solvers.lp(c,A,b)
print(sol['x'])
print(-sol['primal objective'])
print(-numpy.matmul(numpy.transpose(c),sol['x']))

     pcost       dcost       gap    pres   dres   k/t
 0: -9.5935e+01 -5.1217e+02  9e+01  9e-02  2e+00  1e+00
 1: -1.0971e+02 -3.3163e+02  6e+01  5e-02  1e+00  2e+00
 2: -1.2497e+02 -2.6385e+02  5e+01  3e-02  6e-01  2e+00
 3: -1.6339e+02 -1.8688e+02  2e+01  5e-03  1e-01  1e+00
 4: -1.6869e+02 -1.6896e+02  2e-01  6e-05  1e-03  2e-02
 5: -1.6875e+02 -1.6875e+02  2e-03  6e-07  1e-05  2e-04
 6: -1.6875e+02 -1.6875e+02  2e-05  6e-09  1e-07  2e-06
 7: -1.6875e+02 -1.6875e+02  2e-07  6e-11  1e-09  2e-08
Optimal solution found.
[ 6.91e-10]
[ 1.12e+01]

168.74999994212206
[[168.74999994]]


#### *Solution 2*

In [33]:
# try to solve with CVXPY
# plant A
xa1=cvx.Int()
xa2=cvx.Int()

#objective
objective = cvx.Maximize(10*xa1+15*xa2)

#constraint
c1 = 4*xa1+4*xa2<=75
c2 = 4*xa1+2*xa2<=80
c3 = 2*xa1+5*xa2<=60
c4 = 0 <= xa1
c5 = 0 <= xa2
con = [c1,c2,c3,c4,c5]

#solving
prob = cvx.Problem(objective, con)
result = prob.solve()

print ("PLant A should produce", xa1.value, "units of standard product")
print ("Plant A should produce", xa2.value, "units of deluxe product")
print ("The maxmized profit of Plant A is", "$", objective.value)

PLant A should produce 10.000000000492808 units of standard product
Plant A should produce 8.000000003148305 units of deluxe product
The maxmized profit of Plant A is $ 220.00000005215264


In [34]:
# plant B
xb1=cvx.Int()
xb2=cvx.Int()

#objective
objective = cvx.Maximize(10*xb1+15*xb2)

#constraint
c1 = 4*xb1+4*xb2<=45
c2 = 5*xb1+3*xb2<=60
c3 = 5*xb1+6*xb2<=75
c4 = 0 <= xb1
c5 = 0 <= xb2
con = [c1,c2,c3,c4,c5]

#solving
prob = cvx.Problem(objective, con)
result = prob.solve()

print ("PLant B should produce", xb1.value, "units of standard product")
print ("Plant B should produce", xb2.value, "units of deluxe product")
print ("The maxmized profit of Plant B is", "$", objective.value)

PLant B should produce 2.9128592265709495e-09 units of standard product
Plant B should produce 11.000000002512744 units of deluxe product
The maxmized profit of Plant B is $ 165.00000006681975


#### *Solution 3*

In [37]:
# try to solve withg CVXOPT glpk 
from cvxopt import glpk

# Plant A
c=cvxopt.matrix([ -10.0, -15.0],tc='d')
A1=cvxopt.matrix([ [4.0, 4.0, 2.0,-1.0,0.0], [4.0, 2.0, 5.0, 0.0,-1.0] ],tc='d')
A2=cvxopt.matrix([ 75.0, 80.0, 60.0, 0.0, 0.0 ],tc='d')
(status, Ax)=cvxopt.glpk.ilp(c,A1,A2,I=set([0,1]))
profitA= -sum(c.T*Ax)

print ("PLant A should produce", Ax[0], "units of standard product")
print ("Plant A should produce", Ax[1], "units of deluxe product")
print ("The maxmized profit of Plant A is", "$", profitA)

PLant A should produce 10.0 units of standard product
Plant A should produce 8.0 units of deluxe product
The maxmized profit of Plant A is $ 220.0


In [40]:
# Plant B
c=cvxopt.matrix([ -10.0, -15.0],tc='d')
B1=cvxopt.matrix([ [4.0, 4.0, 2.0,-1.0,0.0], [4.0, 2.0, 5.0, 0.0,-1.0] ],tc='d')
B2=cvxopt.matrix([ 45.0, 60.0, 75.0, 0.0, 0.0 ],tc='d')
(status, Bx)=cvxopt.glpk.ilp(c,B1,B2,I=set([0,1]))
profitB= -sum(c.T*Bx)

print ("PLant B should produce", Bx[0], "units of standard product")
print ("Plant B should produce", Bx[1], "units of deluxe product")
print ("The maxmized profit of Plant B is", "$", profitB)

PLant B should produce 0.0 units of standard product
Plant B should produce 11.0 units of deluxe product
The maxmized profit of Plant B is $ 165.0


### 3.	Now, write the optimization model for the joint optimization of Plant A and B, and provide the optimal solution and profit levels. 


---



**Objective Functions:** 

      max 10xa1+15xa2+10xb1+15xb2


**Constraints:** 

*Total Raw Material 120kg; Total Grinding: 80+60=140 ; Total Polishing: 60+75=135*
      
      4xa1+4xa2+4xb1+4b2)<=120
      
      4xa1+2xa2+5xb1+3xb2<=140     
       
      2xa1+5xa2+5xb1+6xb2<=135
      
      xa1,xa2,xb1,xb2>=0



---

The answer is:

PLant A should produce 5.0 units of standard product

Plant A should produce 25.0 units of deluxe product

PLant B should produce -0.0 units of standard product

Plant B should produce -0.0 units of deluxe product

The maxmized profit of Plant A and Plant B is $ 425.0

In [45]:
# Plant A + Plant B Joint Optimization
c=cvxopt.matrix([ -10.0, -15.0, -10.0, -15.0],tc='d')

A=cvxopt.matrix([ [4.0 ,2.0 ,4.0 ,-1.0 ,0.0, 0.0, 0.0], [ 2.0, 5.0 ,4.0 , 0.0, -1.0, 0.0, 0.0], 
                [5.0, 5.0, 4.0, 0.0, 0.0, -1.0, 0.0],[3.0, 6.0, 4.0, 0.0, 0.0, 0.0, -1.0]],tc='d')
b=cvxopt.matrix([140.0, 135.0, 120.0, 0.0, 0.0, 0.0, 0.0 ],tc='d')
(status, Jx)=cvxopt.glpk.ilp(c,A,b,I=set([0,1]))
profitJ= -sum(c.T*Jx)

print ("PLant A should produce", Jx[0], "units of standard product")
print ("Plant A should produce", Jx[1], "units of deluxe product")
print ("PLant B should produce", Jx[2], "units of standard product")
print ("Plant B should produce", Jx[3], "units of deluxe product")
print ("The maxmized profit of Plant A and Plant B is", "$", profitJ)

PLant A should produce 5.0 units of standard product
Plant A should produce 25.0 units of deluxe product
PLant B should produce -0.0 units of standard product
Plant B should produce -0.0 units of deluxe product
The maxmized profit of Plant A and Plant B is $ 425.0


### 4.	Please comment on the results? How do they compare to the results obtained in Question 2.


---
In Question 2, we calculated the maximized profit of plant A and plant B individually. When plant A produces 10 units of standard product and 8 units of deluxe product, it can achieve a maximized profit of 220 dollars. When plant B produces 11units of deluxe product and does not produce any standard product, it can achieve a maximized profit of  165 dollars. Operating separately, plant A and plant together have a maximized profit of 220+165=385 dollars. In Question 3, we conducted a joint optimization of plant A and plant B. When the two plants combinedly plan their production, they can have a maximized profit of 425 dollars, with only plant A producing 5 units of standard product and 25 units of deluxe products.  We can see that, when we do a joint optimization, the total profit is higher than individually optimizing the profit of each plant.

### 5. In one paragraph, detail the recommendations you would provide to the CEO of the company to improve the firm’s operations?

---

My recommendation is to close down plant B and only make production at plant A. According to the joint optimization result, only when plant B doesn’t produce anything and leave all the raw materials for plant A to use its maximum processing capacity, can a joint maximized profit be achieved. Also, we know that plant B has older machines than plant A. Therefore, closing down plant B seems to be a reasonable choice to increase profit. 

## 3.  Hotel La Quinta Motor Inns (LQM)
### Background
LQM is a middle-sized hotel chain that is considering expanding to more locations. LQM used data on 75 existing inn locations to build a linear regression model to predict  “Profitability”, computed at the operating margin, or earnings before interest and taxes divided by total revenue. They tried many independent variables and came up with the final model. All independent variables are significant and were normalized to have mean zero and standard deviation 1. 

     𝑃𝑟𝑜𝑓𝑖𝑡𝑎𝑏𝑖𝑙𝑖𝑡𝑦 = 39.05 − 5.41×𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛 + 5.86×𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡ℎ𝑒 𝐼𝑛𝑛  
                     − 3.09×𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡ℎ𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡ℎ𝑒 𝑎𝑟𝑒𝑎  
                     + 1.75×𝐶𝑜𝑙𝑙𝑒𝑔𝑒 𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠 𝑖𝑛 𝑡ℎ𝑒 𝐴𝑟𝑒𝑎



### 1. According to the regression equation given above, which variable positively affect Profitability? Which variable negatively affect Profitability? Does this intuitively make sense? Why?

---






### 2. Using this regression equation, LQM created a spreadsheet model to predict profitability. LQM collected data for several locations in California, which is provided in the excel spreadsheet on Canvas “LQM”. Using this spreadsheet, compute the profitability for each hotel. Which one has the highest profitability? Which one has the lowest profitability?

---

The answer is:

     Hotel No.2 in Fresno, CA has the highest profitability and hotel No.8 in Los Angeles, CA has the lowest profitability.

In [49]:
!pip install xlrd

    100% |████████████████████████████████| 112kB 2.5MB/s 


In [46]:
# import packages
import pandas as pd
from google.colab import files

# upload the LQM data file to colab
uploaded = files.upload()

Saving LQM.xlsx to LQM.xlsx


In [101]:
# read data into dataframe 
df=pd.read_excel('LQM.xlsx')
df.head()

SELECTING PROFITABLE HOTELS           Unnamed: 1 Unnamed: 2  \
0                          NaN                 NaN        NaN   
1                        Hotel            Location      Price   
2                            1  Eureka, California    2925000   
3                            2  Fresno, California   10000000   
4                            3  Fresno, California    3750000   

           Unnamed: 3                                 Unnamed: 4  \
0                 NaN                                        NaN   
1  Price (normalized)  Square Root of Median Income (normalized)   
2           -0.301823                                   -0.81278   
3             1.69908                                  -0.408199   
4          -0.0685029                                  -0.408199   

                              Unnamed: 5  \
0                                    NaN   
1  College Students in Area (normalized)   
2                              -0.536413   
3                               0.311669   
4                               0.311669   

                              Unnamed: 6  
0                                    NaN  
1  State Population Per Inn (normalized)  
2                              -0.995987  
3                              -0.474279  
4                              -0.474279

In [0]:
# rename columns with features needed in the model
df = df.rename(columns={'SELECTING PROFITABLE HOTELS ':'hotel_id','Unnamed: 1': 'location','Unnamed: 3': 'price', 
                        'Unnamed: 4': 'income', 'Unnamed: 5': 'students', 'Unnamed: 6': 'population','Unnamed: 2':'price_before_normalized' })

In [103]:
# drop unnecessary rows and columns
df=df.drop([0,1], axis=0)
df.head()

hotel_id            location price_before_normalized      price    income  \
2        1  Eureka, California                 2925000  -0.301823  -0.81278   
3        2  Fresno, California                10000000    1.69908 -0.408199   
4        3  Fresno, California                 3750000 -0.0685029 -0.408199   
5        4  Fresno, California                 3500000  -0.139206 -0.408199   
6        5  Fresno, California                  325000   -1.03714 -0.408199   

   students population  
2 -0.536413  -0.995987  
3  0.311669  -0.474279  
4  0.311669  -0.474279  
5  0.311669  -0.474279  
6  0.311669  -0.474279

In [105]:
# Profit = 39.05 − 5.41×𝑆𝑡𝑎𝑡𝑒𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛𝑝𝑒𝑟𝐼𝑛𝑛 + 5.86×𝑃𝑟𝑖𝑐𝑒𝑜𝑓𝑡h𝑒𝐼𝑛𝑛 − 3.09×𝑆𝑞𝑢𝑎𝑟𝑒𝑅𝑜𝑜𝑡𝑜𝑓𝑡h𝑒𝑀𝑒𝑑𝑖𝑎𝑛𝐼𝑛𝑐𝑜𝑚𝑒𝑖𝑛𝑡h𝑒𝑎𝑟𝑒𝑎 + 1.75×𝐶𝑜𝑙𝑙𝑒𝑔𝑒𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠𝑖𝑛𝑡h𝑒𝐴𝑟𝑒𝑎
# calculate profit of each hotel
df_profit = df.copy()
profit = 39.05-5.41*df.population+5.86*df.price-3.09*df.income+1.75*df.students
df_profit['normalized_profit'] = profit

# sort hotel by profit
df_profit.sort_values('normalized_profit',ascending=False)

hotel_id                      location price_before_normalized      price  \
3         2            Fresno, California                10000000    1.69908   
7         6        Long Beach, California                 8950000    1.40212   
2         1            Eureka, California                 2925000  -0.301823   
4         3            Fresno, California                 3750000 -0.0685029   
5         4            Fresno, California                 3500000  -0.139206   
15       14  South Lake Tahoe, California                 3750000 -0.0685029   
13       12  South Lake Tahoe, California                 2500000  -0.422019   
14       13  South Lake Tahoe, California                 1975000  -0.570495   
11       10  South Lake Tahoe, California                 1650000  -0.662409   
16       15  South Lake Tahoe, California                 1475000  -0.711902   
12       11  South Lake Tahoe, California                 1125000  -0.810886   
17       16  South Lake Tahoe, California                  750000  -0.916941   
6         5            Fresno, California                  325000   -1.03714   
10        9       Los Angeles, California                 4900000   0.256732   
8         7       Los Angeles, California                 1950000  -0.577566   
9         8       Los Angeles, California                 1750000  -0.634128   

      income  students population normalized_profit  
3  -0.408199  0.311669  -0.474279           53.3792  
7   0.657845  0.483711  -0.557278           49.0951  
2   -0.81278 -0.536413  -0.995987           44.2424  
4  -0.408199  0.311669  -0.474279           43.0212  
5  -0.408199  0.311669  -0.474279           42.6069  
15 -0.791006 -0.594905  -0.426851            42.361  
13 -0.791006 -0.594905  -0.426851           40.2894  
14 -0.791006 -0.594905  -0.426851           39.4193  
11 -0.791006 -0.594905  -0.426851           38.8807  
16 -0.791006 -0.594905  -0.426851           38.5906  
12 -0.791006 -0.594905  -0.426851           38.0106  
17 -0.791006 -0.594905  -0.426851           37.3891  
6  -0.408199  0.311669  -0.474279            37.345  
10  0.167686   3.10621    3.10653           28.6658  
8   0.167686   3.10621    3.10653           23.7769  
9   0.167686   3.10621    3.10653           23.4454

### 3. LQM has a budget of 10 million to spend on hotels. Suppose we used a “greedy” approach where we selected the most profitable hotels until we ran out of budget. So we would start by the most profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second most profitable, and so on.


---


#### a. Describe what we would do with this approach, i.e., which hotels would we purchase? 

With the greedy approach, we will purchase hotels starting from the top of list and move downwards. When we have 10 million dollars, we can purchase the top 1 hotel on the list, Hotel No.2 in Fresno, CA.

#### b. What would our total predicted profitability be? (This is the sum of the predicted profitability of all hotels we purchase.)

The predicted profitability(before normalization) is: 

39.05 - 5.41*-0.474279 + 5.86*10000000 - 3.09*-0.408199 + 1.75*0.311669 = $58600043.42260505

The predicted profitability(with normalization) is 53.3792

#### c.If we are trying to maximize our total predicted profitability, is this a good approach? How about if we were trying to maximize the average predicted profitability of the hotels we select? How about if we had a budget of 20 million instead of 10 million?

It's not a approach to maximize the total predicted profitability, because the total profitability of a few hotels with lower profitability might be higher than a single hotel even when it has the highest profitability. Trying to maximize the average predicted profitability of the hotels we select is a more reasonable approach in this case.
When our budget increase to 20million, we can try out various combinations of the the two approach to see with which combination it works the best. 


### 4. Now, build an optimization model to select hotels given the 10 million budget. 

---


#### a. Write out the optimization problem. Make sure to detail the decision variables, the objective function and the constraints.

**Decision Variables:** 

Let n=1,2,...,16 reperesents the 16 different hotels, and binary varible Xn represents the choice of buying a hotel ----when Xn = 1, it means LQM should buy this hotel; when Xn = 0, it means LQM shoud not buy the hotel.

**Objective Function:** 

To maximize profitability of selected hotels

\begin{equation}
\text{max}\quad \sum_{n=1}^{16} profitability_{n}\times x_{n}
\end{equation}

**Constraints**: 

A budget of 10 million dollars

\begin{equation}
\quad \sum_{n=1}^{16}  price_{n}\times x_{n}\leq 10,000,000\\
\end{equation}

#### b. What is the optimal solution? Give the values of the decision variables and the optimal objective function value.

In order to maximize the total predicted profitability, we should purchase hotel No.5, No.10, No.11, No.12, No.13, No.15, No.16.

The optimal objective function value is 269.9246813795313.


#### c. Does the optimal solution make sense intuitively? How does it compared to the greedy solution?
Intuitively, the optimal solution makes sense to me. Comparing to the greedy solution's total predicted profitability of 53.3792, this approach can achieve a much higher profitability by purchasing a combination of hotels.

In [99]:
# define decision variables
x=cvx.Bool(16)  
price=np.matrix([2925000,10000000,3750000,3500000,325000,8950000,1950000,1750000,4900000,1650000,1125000,2500000,
                 1975000,3750000,1475000,750000])
profitability=np.matrix([44.24236879,53.37919231,43.02117894,42.6068584,37.34498761,49.09506947,23.77686566,
                         23.44540924,28.66584798,38.88067311,38.01059999,40.28936293,39.41928981,42.36096561,
                         38.59064874,37.38911919])

# define objective
objective=cvx.Maximize(cvx.sum_entries(profitability*x))  

# define constraints
con=[cvx.sum_entries(price*x)<=10000000]  

# solve the problem
prob=cvx.Problem(objective, con)
selection=prob.solve()

print("The optimal profitability of LQM is:", selection)
print("The choice of buying a hotel is", x.value)

The optimal profitability of LQM is: 269.9246813795313
The choice of buying a hotel is [[ 1.09310776e-13]
 [-8.04116515e-14]
 [ 2.21232959e-13]
 [ 2.07350241e-13]
 [ 1.00000000e+00]
 [-3.70481569e-14]
 [ 2.13841339e-13]
 [ 1.62603412e-13]
 [ 1.92789269e-13]
 [ 1.00000000e+00]
 [ 1.00000000e+00]
 [ 1.00000000e+00]
 [ 1.00000000e+00]
 [ 2.24582739e-13]
 [ 1.00000000e+00]
 [ 1.00000000e+00]]


### 5. LQM thinks that buying too many hotels in one city is probably not a good idea and would prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the number of hotels purchased in any city to at most 2.

---

#### a. What are the constraints that you need to add to the model? Intuitively, do you expect the new optimal objective function value to be larger, smaller or the same as before?

**New Constraints Added:**

There are more than multiple hotels in these 3 locations, Fresno, South Lake Tahoe and Los Angeles, so new constraints are needed for them.

\begin{align}
\text{Fresno:}\quad \sum_{n=2}^{5} x_n&\leq 2\\
\text{South Lake Tahoe:}\quad \sum_{n=10}^{16} x_n&\leq 2\\
\text{Los Angeles:}\quad \sum_{n=7}^{9} x_n&\leq 2\\
\end{align} 

Intuitively, I expect the new optimal objective function value to be smaller, since there are more constraints in the model now.

#### b. Write the new optimization model. 

See code below.

#### c. Solve the new model. Give the values of the decision variables and the optimal objective function value. How does this compare to the previous solution?

In order to maximize the total predicted profitability, we should purchase hotel No.1, No.5, No.7, No.8, No.10, No.11.

The optimal objective function value is 205.70090439717904.

Comparing to the previous solution, this solution decreases profitability. A higher budget doesn't always bring higher profitability.

In [100]:
# define decision variables
x=cvx.Bool(16) 

price=np.matrix([2925000,10000000,3750000,3500000,325000,8950000,1950000,1750000,4900000,1650000,1125000,2500000,
                 1975000,3750000,1475000,750000])
profitability=np.matrix([44.24236879,53.37919231,43.02117894,42.6068584,37.34498761,49.09506947,23.77686566,
                         23.44540924,28.66584798,38.88067311,38.01059999,40.28936293,39.41928981,42.36096561,
                         38.59064874,37.38911919])

#  define objective
objective=cvx.Maximize(cvx.sum_entries(profitability*x))  # to maximize the average predicted profitability

# define constraints
c1=cvx.sum_entries(price*x)<=10000000  
## Fresno
c2=x[1]+x[2]+x[3]+x[4]<=2    
## LA
c3=x[6]+x[7]+x[8]<=2 
## SoutH Lake Tahoe
c4=x[9]+x[10]+x[11]+x[12]+x[13]+x[14]+x[15]<=2
con=[c1,c2,c3,c4]

# solve the problem
prob=cvx.Problem(objective, con)
selection=prob.solve()

print("The optimal profitability of LQM is:", selection)
print("The choice of buying a hotel is", x.value)

The optimal profitability of LQM is: 205.70090439717904
The choice of buying a hotel is [[ 1.00000000e+00]
 [ 8.46881348e-12]
 [ 4.96142431e-12]
 [ 4.76555576e-12]
 [ 1.00000000e+00]
 [ 8.07499399e-12]
 [ 1.00000000e+00]
 [ 1.00000000e+00]
 [-8.79486802e-12]
 [ 1.00000000e+00]
 [ 1.00000000e+00]
 [-1.31327298e-11]
 [-1.55206693e-11]
 [-8.42586208e-12]
 [-1.86831456e-11]
 [-3.25040144e-11]]


### 6.	In one paragraph, describe how you would present your results to LQM. Do you have any recommendations for them to improve the regression model? How about to improve the optimization model? 

---

According to optimization result, LQM should invest in a combination of hotels selected from the list and set the budget at 10 million dollars in order to achieve maximized profitability. My recommendation to imprve the optimization model is to conduct sensitive analyses on a reasonable range of budget to see how much budget can bring the highest profitabitiy.
